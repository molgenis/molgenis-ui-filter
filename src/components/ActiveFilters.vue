<template>
  <div>
    <button
      v-for="(item, key) in activeValues"
      :key="key"
      type="button"
      class="active-filter btn btn-light m-1 btn-outline-secondary"
      @click="removeFilter(item)"
    >
      {{ item.label }}: {{ item.value }}
      <font-awesome-icon
        icon="times"
        class="ml-1"
      />
    </button>
  </div>
</template>

<script>
import Vue from 'vue'

import { library } from '@fortawesome/fontawesome-svg-core'
import { faTimes } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
library.add(faTimes)

export default Vue.extend({
  name: 'ActiveFilters',
  components: { FontAwesomeIcon },
  props: {
    filters: {
      type: Array,
      required: true
    },
    value: {
      type: Object,
      default: () => ({})
    }
  },
  data () {
    return {
      activeValues: []
    }
  },
  watch: {
    value: {
      handler (newValue) {
        this.buildActiveValues(newValue)
      },
      immediate: true
    }
  },
  methods: {
    async buildActiveValues (newValue) {
      const activeValues = []
      Object.entries(newValue).forEach(async ([key, current]) => {
        const filter = this.selectFilter(key)

        // Clean op the values by removing undefined entry's
        if (current === undefined || !current.length) return

        // Unpack array
        if (Array.isArray(current)) {
          // Checkbox
          if (filter.type === 'checkbox-filter') {
            // resolve options function/promise and show results later
            const option = await filter.options()
            current.forEach(subKey => {
              const findTextFromValue = option.filter(
                filter => filter.value === subKey
              )[0]
              activeValues.push({
                key,
                subKey,
                value: findTextFromValue.text,
                label: filter.label
              })
            })
          }
          // Range Filter
          if (filter.type === 'range-filter') {
            activeValues.push({
              key,
              value: `${current[0]} to ${current[1]}`,
              label: filter.label
            })
          }
          if (filter.type === 'multi-filter') {
            const options = await filter.options(`?q=id=in=(${current.join(',')})`)
            current.forEach(subKey => {
              const findTextFromValue = options.filter(
                filter => filter.value === subKey
              )[0]
              activeValues.push({
                key,
                subKey,
                value: findTextFromValue.text,
                label: filter.label
              })
            })
          }
        } else {
          // Single item
          activeValues.push({ key, value: current, label: filter.label })
        }
      })
      if (this.value === newValue) {
        this.activeValues = activeValues
      }
    },
    selectFilter (key) {
      return this.filters.filter(filter => filter.name === key)[0]
    },
    removeFilter ({ key, subKey }) {
      if (subKey === undefined) {
        let selections = { ...this.value }
        delete selections[key]
        this.$emit('input', selections)
      } else {
        let selections = { ...this.value }
        selections[key] = selections[key].filter(key => key !== subKey)
        this.$emit('input', selections)
      }
    }
  }
})
</script>

<style scoped>
button svg path {
  transition: fill 0.3s;
}
button:hover svg path {
  fill: var(--danger);
}
</style>
