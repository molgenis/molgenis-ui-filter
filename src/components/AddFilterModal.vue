<template>
  <div>
    <b-button
      v-b-modal.modal-add-filter
      variant="outline-secondary"
      class="w-100 mt-2 add-button"
    >
      Add a filter
      <font-awesome-icon
        class="ml-1"
        icon="plus"
      />
    </b-button>

    <b-modal
      id="modal-add-filter"
      title="Add filter"
      @ok="addFilter"
      @show="selected = inactiveFilters[0].id"
    >
      <b-form-select
        v-model="selected"
        name="filter"
        :options="inactiveFilters"
        value-field="id"
        text-field="name"
      />
    </b-modal>
  </div>
</template>

<script>
import Vue from 'vue'

import { library } from '@fortawesome/fontawesome-svg-core'
import { faPlus } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
library.add(faPlus)

export default Vue.extend({
  name: 'AddFilterModal',
  components: { FontAwesomeIcon },
  store: {
    'filters': 'filters.available'
  },
  data () {
    return {
      selected: null
    }
  },
  computed: {
    inactiveFilters () {
      return this.filters.filter(filter => !filter.active).map((f) => ({
        id: f.id,
        name: f.label
      }))
    }
  },
  methods: {
    addFilter () {
      const filter = this.filters.find((f) => f.id === this.selected)
      filter.active = true
    }
  }
})
</script>
