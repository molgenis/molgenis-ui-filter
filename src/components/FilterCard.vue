<template>
  <b-card
    no-body
    class="filter-card mb-2"
  >
    <b-card-header
      :class="filter.collapsible ? 'collapsable' : ''"
      @click="toggleCollapse"
    >
      <div
        class="title mr-3 px-1"
        :title="filter.label"
      >
        <font-awesome-icon
          v-if="filter.collapsible"
          icon="caret-right"
          :style="iconStyle"
          class="mr-2"
        />
        {{ filter.label }}
        <button
          v-if="canRemove"
          class="remove-button"
          @click.stop="removeFilter"
        >
          <font-awesome-icon icon="times" />
        </button>
      </div>
    </b-card-header>

    <b-collapse
      :id="filter.id"
      :visible="isOpen"
    >
      <b-card-body>
        <b-form-group :description="description">
          <slot />
        </b-form-group>
      </b-card-body>
    </b-collapse>
  </b-card>
</template>

<script>
import Vue from 'vue'

import { library } from '@fortawesome/fontawesome-svg-core'
import { faCaretRight, faTimes } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
library.add(faCaretRight, faTimes)

export default Vue.extend({
  name: 'FilterCard',
  components: { FontAwesomeIcon },
  props: {
    filter: {
      type: Object,
      default: () => {}
    },
    description: {
      type: String,
      required: false,
      default: () => ''
    },
    canRemove: {
      type: Boolean,
      required: false,
      default: () => false
    }
  },
  computed: {
    iconStyle () {
      return {
        transform: `rotate(${this.filter.collapsed ? 90 : 0}deg)`,
        transition: 'transform 0.2s'
      }
    },
    isOpen: function () {
      if (this.filter.collapsible) {
        return this.filter.collapsed
      }

      return true
    }
  },
  methods: {
    removeFilter () {
      this.filter.active = false
    },
    toggleCollapse: function () {
      if (this.filter.collapsible) {
        this.filter.collapsed = !this.filter.collapsed
      }
    }
  }
})
</script>

<style scoped>
  .card-header.collapsable {
    cursor: pointer;
  }
  .form-group {
    margin-bottom:0;
  }
  .title{
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }
  .remove-button{
    background: none;
    border: 0;
    transition: opacity 0.2s, color 0.2s;
    opacity: 0;
    height: inherit;
    width: 1.5em;
    text-align: center;
    display: inline-block;
    position: absolute;
    right: 10px;
    cursor: pointer;
  }
  .remove-button:hover{
    color: var(--danger);
  }
  .filter-card:hover .remove-button{
    opacity: 1;
  }
  .sortable-ghost{
    border-style: dashed;
  }
  .sortable-ghost > div{
    opacity: 0.2;
  }
</style>
