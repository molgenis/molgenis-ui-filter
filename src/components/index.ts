export { default as FilterCard } from './FilterCard.vue'
export { default as FilterContainer } from './FilterContainer.vue'
export { default as AddFilterModal } from './AddFilterModal.vue'
export { default as ActiveFilters } from './ActiveFilters.vue'
