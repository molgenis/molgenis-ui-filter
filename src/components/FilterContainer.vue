<template>
  <div
    class="filter-container"
    @mouseup="drag=false"
  >
    <b-collapse
      id="mobile-button-toggle"
      :visible="doCollapse"
    >
      <button
        class="btn w-100 my-2 btn-outline-secondary"
        @click="mobileToggle=!mobileToggle"
      >
        {{ mobileToggle?'Hide filters':'Show filters' }}
      </button>
    </b-collapse>

    <b-collapse
      id="mobile-toggle"
      :visible="!doCollapse || mobileToggle"
    >
      <draggable
        v-model="filters"
        :disabled="!doDragDrop"
        :class="{'dragdrop': doDragDrop, 'dragging': drag}"
        @choose="drag=true"
        @end="drag=false"
      >
        <transition-group>
          <filter-card
            v-for="filter in visibleFilters"
            :key="filter.id"
            :filter="filter"
            :can-remove="canEdit"
            @removeFilter="filter.visible = false"
          >
            <component
              :is="filter.type"
              :filter="filter"
            />
          </filter-card>
        </transition-group>
      </draggable>

      <add-filter-modal
        v-if="canEdit && invisibleFilters.length > 0"
        v-model="filters.visible"
        :filters="invisibleFilters"
      />
    </b-collapse>
  </div>
</template>

<script>
import AddFilterModal from './AddFilterModal.vue'
import { FilterCard } from '.'
import { StringFilter, CheckboxFilter, NumberFilter, RangeFilter, MultiFilter } from './filters/'
import draggable from 'vuedraggable'

export default {
  name: 'FilterContainer',
  components: { AddFilterModal, StringFilter, CheckboxFilter, FilterCard, NumberFilter, RangeFilter, MultiFilter, draggable },
  props: {
    canEdit: {
      type: Boolean,
      required: false,
      default: () => false
    }
  },
  data () {
    return {
      drag: false,
      width: 0,
      mobileToggle: false
    }
  },
  computed: {
    doCollapse () {
      // Bootstrap's mobile collapse width
      return this.width <= 576
    },
    doDragDrop () {
      return this.canEdit && !this.doCollapse
    },
    invisibleFilters () {
      return this.filters.filter(filter => !filter.visible)
    },
    visibleFilters () {
      return this.filters.filter(filter => filter.visible)
    }
  },
  created () {
    window.addEventListener('resize', this.handleResize)
    this.handleResize()
  },
  destroyed () {
    window.removeEventListener('resize', this.handleResize)
  },
  methods: {
    handleResize () {
      this.width = window.innerWidth
    }
  },
  store: ['filters']
}
</script>

<style scoped>
  .dragdrop{
    cursor: grab;
  }
  .dragdrop.dragging{
    cursor: grabbing;
  }
</style>
