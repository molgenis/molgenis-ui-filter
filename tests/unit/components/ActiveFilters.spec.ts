import { mount } from '@vue/test-utils'
import { ActiveFilters } from '@/components'
import mockData from '../../mockdata'

describe('ActiveFilters.vue', () => {
  const optionsPromise = () => {
    return new Promise(
      function (resolve) {
        resolve(mockData.checkboxLotsOptions)
      }
    )
  }
  const wrapper = mount(ActiveFilters, {
    propsData: {
      value: {
        string: 'blah',
        checkbox: ['red'],
        multi: ['blue', 'purple'],
        range: [-5, 5]
      },
      filters: [ {
        name: 'string',
        label: 'String',
        description: 'search by string',
        placeholder: 'placeholder',
        type: 'string-filter',
        collapsable: false
      }, {
        name: 'checkbox',
        label: 'Checkbox',
        collapsed: false,
        bulkOperation: true,
        options: optionsPromise,
        type: 'checkbox-filter'
      }, {
        name: 'multi',
        label: 'Checkbox',
        collapsed: false,
        bulkOperation: true,
        options: optionsPromise,
        type: 'multi-filter'
      }, {
        name: 'range',
        label: 'Range',
        min: -10,
        max: 10,
        collapsed: false,
        options: optionsPromise,
        type: 'range-filter'
      }]
    }
  })

  it('matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('removes filter values on click', () => {
    // @ts-ignore
    wrapper.vm.removeFilter({ key: 'string' })
    expect(wrapper.emitted().input[0][0]).toEqual({ checkbox: [ 'red' ], 'range': [-5, 5], 'multi': ['blue', 'purple'] })
    // @ts-ignore
    wrapper.vm.removeFilter({ key: 'checkbox', subKey: 'red' })
    expect(wrapper.emitted().input[1][0]).toEqual({ 'checkbox': [], 'range': [-5, 5], 'multi': ['blue', 'purple'], 'string': 'blah' })
  })
})
