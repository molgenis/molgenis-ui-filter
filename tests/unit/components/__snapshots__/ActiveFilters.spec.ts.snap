// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveFilters.vue matches snapshot 1`] = `
<div>
  <button
    class="active-filter btn btn-light m-1 btn-outline-secondary"
    type="button"
  >
    
    String: blah
    
    <svg
      aria-hidden="true"
      class="ml-1 svg-inline--fa fa-times fa-w-11"
      data-icon="times"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 352 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class=""
        d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
        fill="currentColor"
      />
    </svg>
  </button>
  <button
    class="active-filter btn btn-light m-1 btn-outline-secondary"
    type="button"
  >
    
    Range: -5 to 5
    
    <svg
      aria-hidden="true"
      class="ml-1 svg-inline--fa fa-times fa-w-11"
      data-icon="times"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 352 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class=""
        d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
        fill="currentColor"
      />
    </svg>
  </button>
  <button
    class="active-filter btn btn-light m-1 btn-outline-secondary"
    type="button"
  >
    
    Checkbox: Red
    
    <svg
      aria-hidden="true"
      class="ml-1 svg-inline--fa fa-times fa-w-11"
      data-icon="times"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 352 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class=""
        d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
        fill="currentColor"
      />
    </svg>
  </button>
  <button
    class="active-filter btn btn-light m-1 btn-outline-secondary"
    type="button"
  >
    
    Checkbox: Blue
    
    <svg
      aria-hidden="true"
      class="ml-1 svg-inline--fa fa-times fa-w-11"
      data-icon="times"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 352 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class=""
        d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
        fill="currentColor"
      />
    </svg>
  </button>
  <button
    class="active-filter btn btn-light m-1 btn-outline-secondary"
    type="button"
  >
    
    Checkbox: Purple
    
    <svg
      aria-hidden="true"
      class="ml-1 svg-inline--fa fa-times fa-w-11"
      data-icon="times"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 352 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        class=""
        d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
        fill="currentColor"
      />
    </svg>
  </button>
</div>
`;
